projecttacoma/deqm-test-server
# Add practitioner query param functionality to $evaluate-measure

# Summary
This PR implements the practitioner query param for the `$evaluate-measure` operation.

## New behavior
The test server now accepts a `practitioner` ID as a query parameter when invoking the `$evaluate-measure` operation.

According to the `$evaluate-measure` [spec](https://www.hl7.org/fhir/operation-measure-evaluate-measure.html), the practitioner query parameter has cardinality `0..1`. In contrast to the [$care-gaps operation](https://build.fhir.org/ig/HL7/davinci-deqm/OperationDefinition-care-gaps.html), the subject and practitioner query parameters are **not** mutually exclusive for $evaluate-measure. If the practitioner is specified for the measure that will be calculated, the measure will be calculated **only for subjects that have a primary relationship to the identified practitioner**.

## Code changes
Code changes are inspired by [this PR that adds support for the practitioner parameter for the $care-gaps operation](https://github.com/projecttacoma/deqm-test-server/pull/93).

Most code changes appear in `measure.service.js` within `evaluateMeasure()`. Now, the function checks for the presence of the practitioner query param. If present, the function narrows down the patients to calculate the measure against so that the calculation only occurs for patients that reference the practitioner.

Unit tests were added to reflect the new query parameter.

# Testing guidance
Run `npm test` to ensure that the new unit tests pass and that all previous tests still pass. Check that the unit tests cover all possible situations that need to check for a practitioner (ex. A single Patient subject/individual report type, Group subject/population report type, etc.)

Steps for testing the new functionality in the test server:
* Run `npm run db:reset`, followed by `npm run connectathon-upload`
* Send a PUT request to `[http://localhost:3000/4_0_1/Practitioner/1`](http://localhost:3000/4_0_1/Practitioner/1%60) with the following body: 
```
{
	“resourceType”: “Practitioner”,
	“id”: “1”
}
```
* Modify some patients to reference this test practitioner. For example, we can first modify the denominator EXM130 patient to reference this practitioner. Send a PUT request to `[http://localhost:3000/4_0_1/Patient/denom-EXM130`](http://localhost:3000/4_0_1/Patient/denom-EXM130%60) with the following body:
```
{
	"resourceType": "Patient",
	"id": "denom-EXM130",
	“generalPractitioner”: {
		"identifier": {
			"value": “Practitioner/1”
		}
	}
}
```
* Send a GET request to `http://localhost:3000/4_0_1/Measure/measure-EXM130-7.3.000/$evaluate-measure?subject=denom-EXM130&periodStart=2019-01-01&periodEnd=2019-12-31&practitioner=Practitioner/1`. You should receive the typical output for $evaluate-measure with the denominator patient, since the denominator patient references the practitioner
* Now, send a GET request to `http://localhost:3000/4_0_1/Measure/measure-EXM130-7.3.000/$evaluate-measure?subject=numer-EXM130&periodStart=2019-01-01&periodEnd=2019-12-31&practitioner=Practitioner/1`.  Since this subject does not reference the practitioner, the measure will not be calculated for this subject. The output should reflect that.
* Try sending a request that omits the subject query parameter. In this case, in alignment with the spec, we should only have the measure calculated on the denominator patient again since it is the only patient that references our practitioner.
* Try sending a request with a Group. See [this PR](https://github.com/projecttacoma/deqm-test-server/pull/86) for an example Group resource with EXM130 patients. If using this Group, you can send a request to http://localhost:3000/4_0_1/Measure/measure-EXM130-7.3.000/$evaluate-measure?subject=Group/EXM130-patients&periodStart=2019-01-01&periodEnd=2019-12-31&practitioner=Practitioner/1&reportType=population. Since the Group contains the numerator and denominator patients but only the denominator patient references the practitioner, the measure report should reflect this.
* Try sending other requests (have multiple patients reference the practitioner, etc.)
